rollouts/controllers/rollout_controller.go

```diff
@@ -749,13 +794,57 @@ func (r *RolloutReconciler) mapClusterUpdateToRequest(cluster client.Object) []r
 	return requests
 }
 
+func getTargetsForWave(allTargets *Targets,
+	rollout *gitopsv1alpha1.Rollout,
+	waveClusters []gkeclusterapis.ContainerCluster, waveName string) *Targets {
+	targets := &Targets{}
+
+	clusterNames := make(map[string]struct{})
+
+	for _, cluster := range waveClusters {
+		clusterNames[cluster.Name] = struct{}{}
+	}
+
+	for _, toCreate := range allTargets.ToBeCreated {
+		if _, inWave := clusterNames[toCreate.cluster.Name]; inWave {
+			targets.ToBeCreated = append(targets.ToBeCreated, toCreate)
+		}
+	}
+
+	for _, rrs := range allTargets.ToBeUpdated {
+		if _, inWave := clusterNames[rrs.Spec.ClusterRef.Name]; inWave {
+			targets.ToBeUpdated = append(targets.ToBeUpdated, rrs)
+		}
+	}
+
+	for _, rrs := range allTargets.Unchanged {
+		if _, inWave := clusterNames[rrs.Spec.ClusterRef.Name]; inWave {
+			targets.Unchanged = append(targets.Unchanged, rrs)
+		}
+	}
```

Comment:
I wonder if we should remove the targets allocated to waves from allTargets ? (and print the counts at the end, might help us in catching any edge cases)

Reply:
I updated the code so the code only needs to iterate through the targets for each list once (created, updated, unchanged, deleted) assigning each target to its respective wave. This simplifies the code and eliminates the possibility where a target is not associated with a wave.